File: nexus_dev_flutter/lib/models/domain_models.dart
// Domain models ported from the Kotlin Nexus Dev Orchestrator.
// OLED theme uses black/white only; severity is conveyed via white intensity.

enum AgentType {
  planner('Planner Agent', 'Decomposes complex requests into task execution DAG', 'gemini-3.1-pro-preview'),
  architect('Architect Agent', 'System architecture design, module graph & ADR creation', 'gemini-3.1-pro-preview'),
  claudeCode('Claude Code Agent', 'Senior Code Architect & High-Level Refactoring Engine', 'gemini-3.1-pro-preview'),
  codex('Codex Native Agent', 'High-Performance Logic Synthesizer, JNI & NDK Specialist', 'gemini-3.1-pro-preview'),
  exaResearcher('Exa Neural Researcher', 'Deep Web Grounding, GitHub Issues & Tech Intelligence', 'gemini-3.5-flash'),
  debugAgent('Debug & Log Agent', 'Log parser, stacktrace minimization & hypothesis testing', 'gemini-3.1-pro-preview'),
  devsecopsSentinel('DevSecOps Sentinel', 'SAST/DAST Vulnerability Scanner, OWASP & CWE Auditor', 'gemini-3.1-pro-preview'),
  performanceAgent('Performance Agent', 'Memory allocation, GC pauses, DEX count & SIMD speedup', 'gemini-3.5-flash'),
  testAgent('Test Automator', 'Unit, Robolectric JVM, regression & property-based tests', 'gemini-3.5-flash'),
  reviewerAgent('Reviewer Agent', 'Diff inspector, regression blocker & code standards', 'gemini-3.1-pro-preview'),
  releaseAgent('Release Agent', 'Changelog generator, versioning & CI/CD deployment', 'gemini-3.5-flash'),
  nativeNdkEngineer('Native NDK Engineer', 'C++20/Rust/JNI Bindings & CMake Build System Architect', 'gemini-3.5-flash');

  const AgentType(this.displayName, this.roleDescription, this.defaultModel);
  final String displayName;
  final String roleDescription;
  final String defaultModel;

  String get title => displayName;
}

enum AgentState {
  idle,
  planning,
  researching,
  architecting,
  writingCode,
  auditing,
  compiling,
  testing,
  profiling,
  reviewing,
  deploying,
  completed,
  error;

  String get label => name.toUpperCase();
}

class AgentLiveStatus {
  final AgentType agentType;
  final AgentState state;
  final String currentTask;
  final double progress;
  final String thoughts;
  final int activeTokens;
  final int latencyMs;

  const AgentLiveStatus({
    required this.agentType,
    this.state = AgentState.idle,
    this.currentTask = 'Standby',
    this.progress = 0,
    this.thoughts = '',
    this.activeTokens = 0,
    this.latencyMs = 42,
  });

  AgentLiveStatus copyWith({
    AgentState? state,
    String? currentTask,
    String? thoughts,
  }) =>
      AgentLiveStatus(
        agentType: agentType,
        state: state ?? this.state,
        currentTask: currentTask ?? this.currentTask,
        thoughts: thoughts ?? this.thoughts,
        progress: progress,
        activeTokens: activeTokens,
        latencyMs: latencyMs,
      );
}

enum VulnerabilitySeverity { critical, high, medium, low, clean }

class SecurityFinding {
  final String id;
  final String title;
  final VulnerabilitySeverity severity;
  final String cwe;
  final String masvsRef;
  final String masvsCategory;
  final String component;
  final int line;
  final String description;
  final String recommendation;
  final String originalSnippet;
  final String patchSnippet;
  final double confidence;
  final bool autoFixAvailable;

  const SecurityFinding({
    required this.id,
    required this.title,
    required this.severity,
    this.cwe = '',
    this.masvsRef = '',
    this.masvsCategory = '',
    this.component = '',
    this.line = 0,
    this.description = '',
    this.recommendation = '',
    this.originalSnippet = '',
    this.patchSnippet = '',
    this.confidence = 0.95,
    this.autoFixAvailable = true,
  });
}

class MasvsCategoryMetric {
  final String categoryCode;
  final String categoryName;
  final int findingsCount;
  final VulnerabilitySeverity maxSeverity;
  final bool isCompliant;

  const MasvsCategoryMetric({
    required this.categoryCode,
    required this.categoryName,
    required this.findingsCount,
    required this.maxSeverity,
    required this.isCompliant,
  });
}

class MasvsScanResult {
  final String scanId;
  final String targetName;
  final int linesScanned;
  final int complianceScore;
  final String complianceGrade;
  final int totalVulnerabilities;
  final int criticalCount;
  final int highCount;
  final int mediumCount;
  final int lowCount;
  final List<SecurityFinding> findings;
  final List<MasvsCategoryMetric> categories;
  final int scanDurationMs;
  final int autoFixPatchesCount;

  const MasvsScanResult({
    required this.scanId,
    required this.targetName,
    required this.linesScanned,
    required this.complianceScore,
    required this.complianceGrade,
    required this.totalVulnerabilities,
    required this.criticalCount,
    required this.highCount,
    required this.mediumCount,
    required this.lowCount,
    required this.findings,
    required this.categories,
    required this.scanDurationMs,
    required this.autoFixPatchesCount,
  });
}

class TelemetrySnapshot {
  final double cpuPercent;
  final double memoryUsageMb;
  final int activeThreads;
  final int latencyMs;
  final int totalRequests;
  final int totalTokens;
  final double sessionCostUsd;

  const TelemetrySnapshot({
    this.cpuPercent = 12.4,
    this.memoryUsageMb = 94.2,
    this.activeThreads = 8,
    this.latencyMs = 48,
    this.totalRequests = 24,
    this.totalTokens = 15200,
    this.sessionCostUsd = 0.0024,
  });

  TelemetrySnapshot copyWith({
    double? cpuPercent,
    double? memoryUsageMb,
    int? activeThreads,
    int? latencyMs,
    int? totalRequests,
    int? totalTokens,
    double? sessionCostUsd,
  }) =>
      TelemetrySnapshot(
        cpuPercent: cpuPercent ?? this.cpuPercent,
        memoryUsageMb: memoryUsageMb ?? this.memoryUsageMb,
        activeThreads: activeThreads ?? this.activeThreads,
        latencyMs: latencyMs ?? this.latencyMs,
        totalRequests: totalRequests ?? this.totalRequests,
        totalTokens: totalTokens ?? this.totalTokens,
        sessionCostUsd: sessionCostUsd ?? this.sessionCostUsd,
      );
}

enum TerminalShellType { zsh, nushell, fish }

class AutopilotProgress {
  final int stepIndex;
  final int totalSteps;
  final String currentStep;
  final bool active;
  final bool buildPass;
  final bool testsPass;
  final bool securityAcceptable;
  final bool lintPass;
  final int criticalBugsCount;
  final int regressionsCount;
  final List<String> logs;

  const AutopilotProgress({
    this.stepIndex = 0,
    this.totalSteps = 12,
    this.currentStep = 'Standby',
    this.active = false,
    this.buildPass = true,
    this.testsPass = true,
    this.securityAcceptable = true,
    this.lintPass = true,
    this.criticalBugsCount = 0,
    this.regressionsCount = 0,
    this.logs = const [],
  });

  AutopilotProgress copyWith({
    int? stepIndex,
    int? totalSteps,
    String? currentStep,
    bool? active,
    bool? buildPass,
    bool? testsPass,
    bool? securityAcceptable,
    bool? lintPass,
    int? criticalBugsCount,
    int? regressionsCount,
    List<String>? logs,
  }) =>
      AutopilotProgress(
        stepIndex: stepIndex ?? this.stepIndex,
        totalSteps: totalSteps ?? this.totalSteps,
        currentStep: currentStep ?? this.currentStep,
        active: active ?? this.active,
        buildPass: buildPass ?? this.buildPass,
        testsPass: testsPass ?? this.testsPass,
        securityAcceptable: securityAcceptable ?? this.securityAcceptable,
        lintPass: lintPass ?? this.lintPass,
        criticalBugsCount: criticalBugsCount ?? this.criticalBugsCount,
        regressionsCount: regressionsCount ?? this.regressionsCount,
        logs: logs ?? this.logs,
      );
}

class NativeModuleSpec {
  final String moduleName;
  final String language;
  final String packageName;
  final List<String> targetArch;
  final List<String> jniMethods;
  final String cppSource;
  final String headerSource;
  final String cmakeListsContent;
  final String kotlinWrapperCode;
  final bool isCompiled;

  const NativeModuleSpec({
    required this.moduleName,
    this.language = 'C++20',
    this.packageName = 'com.example.nativemodule',
    this.targetArch = const ['arm64-v8a', 'armeabi-v7a', 'x86_64'],
    this.jniMethods = const [],
    this.cppSource = '',
    this.headerSource = '',
    this.cmakeListsContent = '',
    this.kotlinWrapperCode = '',
    this.isCompiled = false,
  });
}

class PerformanceMetricsProfile {
  final double cpuUtilizationPercent;
  final double memoryUsageMb;
  final double apkSizeMb;
  final int dexMethodCount;
  final int coldStartTimeMs;
  final int warmStartTimeMs;
  final double gcPauseAverageMs;
  final double jniBridgeLatencyMicroseconds;
  final double nativeSimdSpeedup;

  const PerformanceMetricsProfile({
    this.cpuUtilizationPercent = 14.2,
    this.memoryUsageMb = 88.5,
    this.apkSizeMb = 18.4,
    this.dexMethodCount = 34210,
    this.coldStartTimeMs = 310,
    this.warmStartTimeMs = 85,
    this.gcPauseAverageMs = 2.1,
    this.jniBridgeLatencyMicroseconds = 0.45,
    this.nativeSimdSpeedup = 4.8,
  });
}

class PipelineStageInfo {
  final String name;
  final String status;
  final int durationSeconds;
  final String logSummary;

  const PipelineStageInfo({
    required this.name,
    required this.status,
    required this.durationSeconds,
    this.logSummary = '',
  });
}

class AgentDebateRound {
  final int roundNumber;
  final String claudeArgument;
  final String codexArgument;
  final String evaluatorVerdict;
  final double convergenceScore;

  const AgentDebateRound({
    required this.roundNumber,
    required this.claudeArgument,
    required this.codexArgument,
    required this.evaluatorVerdict,
    required this.convergenceScore,
  });
}

File: app/src/main/java/com/example/domain/model/DomainModels.kt
package com.example.domain.model

enum class AgentType(
    val displayName: String,
    val roleDescription: String,
    val defaultModel: String,
    val badgeColorHex: Long
) {
    PLANNER(
        "Planner Agent",
        "Decomposes complex requests into task execution DAG",
        "gemini-3.1-pro-preview",
        0xFF38BDF8
    ),
    ARCHITECT(
        "Architect Agent",
        "System architecture design, module graph & ADR creation",
        "gemini-3.1-pro-preview",
        0xFF818CF8
    ),
    CLAUDE_CODE(
        "Claude Code Agent",
        "Senior Code Architect & High-Level Refactoring Engine",
        "gemini-3.1-pro-preview",
        0xFF8B5CF6
    ),
    CODEX(
        "Codex Native Agent",
        "High-Performance Logic Synthesizer, JNI & NDK Specialist",
        "gemini-3.1-pro-preview",
        0xFF00F0FF
    ),
    EXA_RESEARCHER(
        "Exa Neural Researcher",
        "Deep Web Grounding, GitHub Issues & Tech Intelligence",
        "gemini-3.5-flash",
        0xFFF59E0B
    ),
    DEBUG_AGENT(
        "Debug & Log Agent",
        "Log parser, stacktrace minimization & hypothesis testing",
        "gemini-3.1-pro-preview",
        0xFFEC4899
    ),
    DEVSECOPS_SENTINEL(
        "DevSecOps Sentinel",
        "SAST/DAST Vulnerability Scanner, OWASP & CWE Auditor",
        "gemini-3.1-pro-preview",
        0xFFEF4444
    ),
    PERFORMANCE_AGENT(
        "Performance Agent",
        "Memory allocation, GC pauses, DEX count & SIMD speedup",
        "gemini-3.5-flash",
        0xFF10B981
    ),
    TEST_AGENT(
        "Test Automator",
        "Unit, Robolectric JVM, regression & property-based tests",
        "gemini-3.5-flash",
        0xFF06B6D4
    ),
    REVIEWER_AGENT(
        "Reviewer Agent",
        "Diff inspector, regression blocker & code standards",
        "gemini-3.1-pro-preview",
        0xFFA855F7
    ),
    RELEASE_AGENT(
        "Release Agent",
        "Changelog generator, versioning & CI/CD deployment",
        "gemini-3.5-flash",
        0xFF6366F1
    ),
    NATIVE_NDK_ENGINEER(
        "Native NDK Engineer",
        "C++20/Rust/JNI Bindings & CMake Build System Architect",
        "gemini-3.5-flash",
        0xFF10B981
    );

    val title: String get() = displayName
}

enum class AgentState {
    IDLE,
    PLANNING,
    RESEARCHING,
    ARCHITECTING,
    WRITING_CODE,
    AUDITING,
    COMPILING,
    TESTING,
    PROFILING,
    REVIEWING,
    DEPLOYING,
    COMPLETED,
    ERROR
}

data class AgentLiveStatus(
    val agentType: AgentType,
    val state: AgentState = AgentState.IDLE,
    val currentTask: String = "Standby",
    val progress: Float = 0f,
    val thoughts: String = "",
    val activeTokens: Int = 0,
    val latencyMs: Long = 42
)

enum class VulnerabilitySeverity(val label: String, val colorHex: Long) {
    CRITICAL("CRITICAL", 0xFFDC2626),
    HIGH("HIGH", 0xFFEA580C),
    MEDIUM("MEDIUM", 0xFFF59E0B),
    LOW("LOW", 0xFF3B82F6),
    CLEAN("CLEAN", 0xFF10B981)
}

data class SecurityFinding(
    val id: String,
    val title: String,
    val severity: VulnerabilitySeverity,
    val cwe: String,
    val masvsRef: String = "",
    val masvsCategory: String = "",
    val component: String = "",
    val line: Int = 0,
    val lineNumber: Int = line,
    val filePath: String = component,
    val description: String = "",
    val recommendation: String = "",
    val originalSnippet: String = "",
    val vulnerableCodeSnippet: String = originalSnippet,
    val patchSnippet: String = "",
    val remediationPatch: String = patchSnippet,
    val confidence: Float = 0.95f,
    val autoFixAvailable: Boolean = true
)

data class SbomPackage(
    val name: String,
    val version: String,
    val purl: String,
    val license: String,
    val sha256Hash: String,
    val vulnerabilitiesCount: Int
)

data class SbomComponent(
    val name: String,
    val version: String,
    val purl: String,
    val license: String,
    val hashSha256: String,
    val vulnerabilitiesCount: Int
)

data class AutopilotPhaseStatus(
    val phaseNumber: Int,
    val phaseName: String,
    val assignedAgent: AgentType,
    val status: PhaseExecutionStatus,
    val details: String = "",
    val durationMs: Long = 0
)

data class AutopilotProgress(
    val stepIndex: Int = 0,
    val totalSteps: Int = 12,
    val currentStep: String = "Standby",
    val active: Boolean = false,
    val buildPass: Boolean = true,
    val testsPass: Boolean = true,
    val securityAcceptable: Boolean = true,
    val lintPass: Boolean = true,
    val criticalBugsCount: Int = 0,
    val regressionsCount: Int = 0,
    val logs: List<String> = emptyList()
)

enum class PhaseExecutionStatus {
    PENDING,
    IN_PROGRESS,
    PASSED,
    FAILED,
    SKIPPED
}

data class NativeModuleSpec(
    val moduleName: String,
    val language: String, // "C++20" or "Rust"
    val packageName: String = "com.example.nativemodule",
    val targetArch: List<String> = listOf("arm64-v8a", "armeabi-v7a", "x86_64"),
    val jniMethods: List<String> = emptyList(),
    val cppSource: String = "",
    val headerSource: String = "",
    val cmakeListsContent: String = "",
    val kotlinWrapperCode: String = "",
    val headerContent: String = headerSource,
    val implementationContent: String = cppSource,
    val cmakeContent: String = cmakeListsContent,
    val jniKotlinWrapper: String = kotlinWrapperCode,
    val isCompiled: Boolean = false
)

data class ProjectBrainNode(
    val id: String,
    val title: String,
    val category: String, // "ADR", "RFC", "CONVERGENCE", "POLICY"
    val status: String,   // "ACCEPTED", "PROPOSED", "SUPERSEDED"
    val content: String,
    val lastUpdated: String
)

data class AgentDebateRound(
    val roundNumber: Int,
    val claudeArgument: String,
    val codexArgument: String,
    val evaluatorVerdict: String,
    val convergenceScore: Float
)

data class PerformanceMetricsProfile(
    val cpuUtilizationPercent: Float = 14.2f,
    val memoryUsageMb: Float = 88.5f,
    val apkSizeMb: Float = 18.4f,
    val dexMethodCount: Int = 34210,
    val coldStartTimeMs: Int = 310,
    val warmStartTimeMs: Int = 85,
    val gcPauseAverageMs: Float = 2.1f,
    val jniBridgeLatencyMicroseconds: Float = 0.45f,
    val nativeSimdSpeedup: Float = 4.8f,
    val recommendations: List<String> = listOf(
        "Enable R8 full mode with shrinkResources to shave ~2.3MB from release APK",
        "Batch native JNI calls in NdkMathAccelerator to reduce JNI boundary transitions",
        "Use Baseline Profiles for Compose startup optimization (<180ms cold start)"
    )
)

data class PipelineStageInfo(
    val name: String,
    val status: String, // "PENDING", "RUNNING", "SUCCESS", "FAILED"
    val durationSeconds: Int,
    val logSummary: String = ""
)

data class TelemetrySnapshot(
    val cpuPercent: Float = 12.4f,
    val memoryUsageMb: Float = 94.2f,
    val activeThreads: Int = 8,
    val latencyMs: Long = 48,
    val totalRequests: Int = 24,
    val totalTokens: Int = 15200,
    val sessionCostUsd: Float = 0.0024f
)

enum class TerminalShellType(
    val shellName: String,
    val promptPrefix: String,
    val promptBadge: String,
    val description: String,
    val tagColorHex: Long
) {
    ZSH(
        shellName = "zsh",
        promptPrefix = "zsh ❯",
        promptBadge = "ZSH",
        description = "Default shell • Powerlevel10k & DevOps POSIX workflows",
        tagColorHex = 0xFF00F0FF
    ),
    NUSHELL(
        shellName = "nushell",
        promptPrefix = "nu ❯",
        promptBadge = "NUSHELL",
        description = "Data exploration • Cloud automation & config manipulation",
        tagColorHex = 0xFF10B981
    ),
    FISH(
        shellName = "fish",
        promptPrefix = "fish ❯",
        promptBadge = "FISH",
        description = "Interactive shell • Instant autosuggestions & syntax highlighting",
        tagColorHex = 0xFFF59E0B
    )
}

// --- Agent Orchestration & Concurrency Models ---
enum class OrchestrationExecutionState {
    IDLE,
    PLANNING_DAG,
    RUNNING_CONCURRENT,
    BARRIER_SYNC,
    EVALUATING,
    COMPLETED,
    FAILED
}

data class ConcurrentAgentTask(
    val taskId: String,
    val agentType: AgentType,
    val title: String,
    val description: String,
    val state: AgentState = AgentState.IDLE,
    val progress: Float = 0f,
    val thoughtsStream: String = "",
    val outputSnippet: String = "",
    val executionTimeMs: Long = 0,
    val isCompleted: Boolean = false,
    val error: String? = null
)

data class SynchronizedAgentOutput(
    val syncId: String,
    val goal: String,
    val claudeArchitectureOutput: String,
    val codexNativeOutput: String,
    val unifiedCodebaseSnapshot: String,
    val conflictCount: Int = 0,
    val testSuitePassed: Boolean = true,
    val convergenceScore: Float = 0.98f,
    val timestamp: Long = System.currentTimeMillis()
)

data class AgentTimelineEvent(
    val id: String,
    val agentType: AgentType,
    val eventType: String,
    val message: String,
    val timestamp: Long = System.currentTimeMillis()
)

// --- MASVS Static Analysis Result Models ---
data class MasvsCategoryMetric(
    val categoryCode: String,      // e.g. "MASVS-STORAGE", "MASVS-CRYPTO"
    val categoryName: String,      // e.g. "Storage Security", "Cryptography"
    val findingsCount: Int,
    val maxSeverity: VulnerabilitySeverity,
    val isCompliant: Boolean
)

data class MasvsScanResult(
    val scanId: String,
    val targetName: String,
    val timestamp: Long = System.currentTimeMillis(),
    val linesScanned: Int,
    val complianceScore: Int,      // 0 - 100
    val complianceGrade: String,   // A+, A, B, C, F
    val totalVulnerabilities: Int,
    val criticalCount: Int,
    val highCount: Int,
    val mediumCount: Int,
    val lowCount: Int,
    val findings: List<SecurityFinding>,
    val categories: List<MasvsCategoryMetric>,
    val scanDurationMs: Long,
    val autoFixPatchesCount: Int
)


